### ISSUE TITLE
Doesn't render react components

### ISSUE THREAD

**Issue by gpad** [keathley/wallaby#176]
Hi to all,
I have a strange issue. I create a new phoenix project and when I try to test with Wallaby, the react components are not rendered while they are easily rendered when I get the page from the browser.

The react components are not directly rendered by my app.js code but are rendered by external code. The cooed seems executed because during the test the log are written in the console 

To reproduce the problem create and empty phoenix app and put this code in `web/templates/layout/app.html.eex`

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <title>test</title>
    <link rel="stylesheet" href="<%= static_path(@conn, "/css/app.css") %>">
    <script src="https://cdnjs.cloudflare.com/ajax/libs/babel-polyfill/6.16.0/polyfill.min.js"></script>
    <link rel="stylesheet" href="https://cdn.getorbita.io/orbita_components.css">
    <script src="https://cdn.getorbita.io/orbita_components.js"></script>
  </head>

  <body>

      <div id="orbita-toolbar-wrapper" class="header-orbita">
        test
      </div>

    </div> <!-- /container -->
    <script src="<%= static_path(@conn, "/js/app.js") %>"></script>
  </body>
</html>
```

In `priv/static/js/app.js`:

```js
const orbita = orbitaComponents.init({
  unauthorizedCallback: function () {},

  toolbar: {
    wrapperId: 'orbita-toolbar-wrapper',
  },
  services: {},
});
```

and create a test like this :

```elixir
defmodule OrbitaHealthLoginTest do
  use OrbitaHealth.AcceptanceCase, async: true

  test "on root I have orbita toolbar", %{session: session} do
    session = session |> visit("/")
    text = session |> find(Query.css(".orbita-components-toolbar")) |> Element.text || ""
    # text = session |> find(Query.css(".header-orbita")) |> Element.text || ""
    assert text != ""
  end
end
```

I configured my app to work with Wallaby.

The commented code that find the element presents in html works fine instead the code not commented it doesn't work.

Is it a bug?

**Comment by PragTob** [comment]
Hi there,

that's hard to see from over here :)

Regarding the app, if you have it already it'd be very helpful if you put the example app in a github repo or gist together with the wallaby setup so we can have a look at it without further overhead :)

It could be a big, it could also be something else - does the component show up rendered on the screenshot when you do `take_screenshot`?

One of the most likely reasons is that `phantomJS` is based on a relatively old webkit browser and might not support some newer JS/browser APIs that you or your browser doesn't support yet (although I see you include the babel polyfill).

**Comment by keathley** [comment]
I can reproduce this locally. I added an `console.log('Unauthorized') to the `unauthorizedCallback` callback. This is what I see in the console:

```
ENVIRONMENT:  production
Try create orbita components ...
>>> IFRAME ERROR:  oauthUrl: 'undefined'
      currentDomain: 'http://localhost:57003'
      and applicationUID: undefined can't be undefined
Unauthorized
```

That would tell me that its at least loading some of the the javascript from the cdn. There's no js error being thrown in the console. Babel polyfill should take care of all your browser inconsistencies. Without knowing more about the internals of your application its hard to say much more.

**Comment by gpad** [comment]
@PragTob This the repo https://github.com/gpad/orbita_health on branch `wallby` you find the code.

I took the screenshot and the component is not rendered.

I also put some code in this function:

```js
const orbita = orbitaComponents.init({
  unauthorizedCallback: function () {
     // some code with Jquery
},

```

To find some DOM nodes and they are presents on firefox but not on phantomJS ...

**Comment by gpad** [comment]
@keathley you can find the app [here](https://github.com/gpad/orbita_health)  the error is normal because I didn't configure all the react component but if run the example you'll see that is rendered correctly.

**Comment by keathley** [comment]
Thanks! I saw the error in the Chrome console as well. I just mean that at least *some* javascript is running. So its _probably_ not an issue with loading the assets from the cdn.

**Comment by PragTob** [comment]
Imo the babel polyfill doesn't take care of everything, e.g. a `NodeList` in my current browser supports `forEach` but with phantom it would always thrown an error, so I had to add `Array.from` (from the polyfill) - so it could be :)

@gpad thanks for making the code available! :)

**Comment by gpad** [comment]
@PragTob @keathley thanks for your help. Do you think that the problem is inside phantom.js? I'm running version 2.1 so I think it's the last, and I think that also the tests of the react component are running with phantom.

Tomorrow I'll try with a simpler react components to verify if there is some _"strange"_ code in the component.

**Comment by keathley** [comment]
I suspect that your issue is related to some incompatibility between phantomjs and the JS that you're trying to run. It _might_ be solved with the correct polyfill. The only other thing that I can think of is that there's potentially an issue with loading the assets over the cdn. But I don't think a cdn issue is that likely since its running the `orbitaComponents.init` function which means phantom was able to load those assets from the cdn.

**Comment by gpad** [comment]
> I suspect that your issue is related to some incompatibility between phantomjs and the JS that you're
> trying to run.

I agree with you @keathley. I'm busy in this days but I would try with a simpler component to verify if the issue is still present.

I'm also one of the developers of orbitaComponents and the code is tested with standard react stack and now it doesn't use phantom but in the old versions I find a reference to [phantomjs-polyfill](https://www.npmjs.com/package/phantomjs-polyfill), a clue not an evidence :wink:

**Comment by keathley** [comment]
@gpad I'm going to close this for now since I'm not sure how else to diagnose the issue. If you discover the root cause feel free to re-open this. Thanks :+1:

**Comment by juddey** [comment]
Hello! I know this is a closed issue but I thought I would share what I found this morning wrestling with react components and phantomjs and wallaby: 

**tldr**; Update `phantomjs` to the latest development build (2.5.0), available here: https://bitbucket.org/ariya/phantomjs/downloads/

Longer: 
I had similiar issues as gpad with my components not rendering. Tried several things (including installing hound, using that with phantomjs, configuring selenium webdriver, using that with hound). I also found that other people have had similar issues with their react components not rendering in phantomjs: https://stackoverflow.com/questions/38469005/why-i-am-not-able-to-render-my-reactjs-application-using-phantomjs-2-1-1

I also found npm's version (the latest stable of phantomjs - 2.1.1) doesn't support ES6 by default, whereas 2.5 includes ES6.

Happy to put in a PR to the readme (or something) given the last stable version of `phantomjs` was at the start of last year.

**Comment by PragTob** [comment]
@juddey thanks for the work digging in and letting us know. 

As a side note, I'm not sure that ES6 is a feature as lots of browsers don't support ES6 yet so transpiling is still a must imo :)

**Comment by juddey** [comment]
Yep, I wasn't questioning transpiling. If you're lucky (like me) you won't need to support older browsers so this might be all you need. I'll leave this issue as is and won't PR into the readme - it'll just have to be taken out again at a later state.
:wave:
